File: Sample-ObjC/06-query-data/Mesosfer-ObjC/QueryUserViewController.m
/**
 * Copyright (c) 2016, Mesosfer
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "QueryUserViewController.h"
#import "UIViewController+Helper.h"
#import <Mesosfer/Mesosfer.h>

@interface QueryUserViewController () <UISearchBarDelegate>

@property (nullable, nonatomic, strong) NSArray<MFUser *>* users;

@end

@implementation QueryUserViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.estimatedRowHeight = 100;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self queryUserWithUsername:nil];
}

- (void)queryUserWithUsername:(NSString*)username {
    MFQuery *query = [MFUser query];
    if (username) {
        [query whereKey:@"firstname" equalTo:username];
    }
    [query findAsyncWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
        if (error) {
            [self showErrorWithTitle:@"Query user failed" error:error handler:nil];
            return;
        }
        
        if (objects) {
            self.users = objects;
            [self.tableView reloadData];
        }
    }];
}

#pragma mark - UITableViewDelegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (self.users) {
        return self.users.count;
    }
    
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    if (self.users) {
        return self.users[section].objectId;
    }
    return nil;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellUser" forIndexPath:indexPath];
    
    if (self.users) {
        MFUser *user = self.users[indexPath.section];
        
        if ([user dictionary]) {
            cell.textLabel.text = [NSString stringWithFormat:@"%@", [user dictionary]];
        }
    }
    
    return cell;
}

#pragma mark - UISearchBarDelegate

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    searchBar.showsCancelButton = YES;
    [self queryUserWithUsername:searchBar.text];
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    searchBar.showsCancelButton = NO;
    searchBar.text = nil;
    [self queryUserWithUsername:nil];
}

@end

File: Sample-Swift/06-query-data/Mesosfer-Swift/QueryUserViewController.swift
/**
 * Copyright (c) 2016, Mesosfer
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import UIKit
import Mesosfer

class QueryUserViewController: UITableViewController {
    
    var users: [MFUser]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.estimatedRowHeight = 100
        self.tableView.rowHeight = UITableViewAutomaticDimension
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        self.queryUser()
    }
    
    func queryUser(firstname: String? = nil) {
        let query = MFUser.query()
        if let firstname = firstname {
            query.whereKey("firstname", equalTo: firstname)
        }
        query.findAsync { (users, error) in
            if let e = error as? NSError {
                self.showError(title: "Query user failed", error: e)
                return
            }
            
            if let users = users as? [MFUser] {
                self.users = users
                self.tableView.reloadData()
            }
        }
    }
}

extension QueryUserViewController {
    override func numberOfSections(in tableView: UITableView) -> Int {
        if let users = self.users {
            return users.count
        }
        return 0
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if let users = self.users {
            return users[section].objectId
        }
        return nil
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellUser", for: indexPath)
        
        if let users = self.users {
            let user = users[indexPath.section]
            
            if let dict = user.dictionary() as? [String:AnyObject] {
                cell.textLabel?.text = "\(dict)"
            }
        }
        
        return cell
    }
}

extension QueryUserViewController: UISearchBarDelegate {
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.showsCancelButton = true
        self.queryUser(firstname: searchBar.text)
    }
    
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        searchBar.showsCancelButton = false
        searchBar.text = nil
        self.queryUser()
    }
}
